### PR TITLE
[BEAM-6611] BigQuery file loads in Streaming for Python SDK

### PR DESCRIPTION
**Pull request by ttanay**
Added support for writing to BigQuery via file loads for the Python SDK.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python3_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) <br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by ttanay** [comment]
Run Python PostCommit

**Comment by ttanay** [comment]
Run Python PostCommit

**Comment by ttanay** [comment]
Run Portable_Python PreCommit

**Comment by ttanay** [comment]
Run Python PostCommit

**Comment by ttanay** [comment]
Run Python PostCommit

**Comment by ttanay** [comment]
Run Python PostCommit

**Comment by ttanay** [comment]
Run Python PostCommit

**Comment by ttanay** [comment]
Run Python PostCommit

**Comment by ttanay** [comment]
This depends on https://github.com/apache/beam/pull/8934.
This makes use of a BigQuery matcher that  polls BQ for results, that was introduced in https://github.com/apache/beam/pull/8934.

**Comment by ttanay** [comment]
Run Python PostCommit

**Comment by ttanay** [comment]
Run Python PreCommit

**Comment by ttanay** [comment]
Run Python PreCommit

**Comment by ttanay** [comment]
Run Python PreCommit

**Comment by ttanay** [comment]
Run Python PostCommit

**Comment by ttanay** [comment]
Run Python PreCommit

**Comment by ttanay** [comment]
Run Python PreCommit

**Comment by ttanay** [comment]
Run Python PostCommit

**Comment by ttanay** [comment]
Run Python PostCommit

**Comment by ttanay** [comment]
R: @pabloem 

The test failure is because the tests:
 - `test_multiple_destinations_transform` (apache_beam.io.gcp.bigquery_test.BigQueryStreamingInsertTransformIntegrationTests) [1] 
 - `test_value_provider_transform` (apache_beam.io.gcp.bigquery_test.BigQueryStreamingInsertTransformIntegrationTests) [2] 

are running in streaming in `:sdks:python:directRunnerIT`[3] even though they were not specified there. 
The only test of that class that was specified was `test_multiple_destinations_transform_streaming`. This does run and pass.

That test - `apache_beam.io.gcp.bigquery_test.BigQueryStreamingInsertTransformIntegrationTests.test_multiple_destinations_transform_streaming`[4] is also on the master branch in the same Gradle task, but, the tests run fine and pass for the master branch.

This Gradle task executes the `run_integration_test.sh` bash script.
I have tried executing the same script with the same params as those that the Gradle task does, but it doesn't run these tests.

I think I'll need help to make this test pass. 

[1] https://github.com/apache/beam/blob/master/sdks/python/apache_beam/io/gcp/bigquery_test.py#L562-L623
[2] https://github.com/apache/beam/blob/master/sdks/python/apache_beam/io/gcp/bigquery_test.py#L501-L560
[3] https://github.com/apache/beam/blob/master/sdks/python/build.gradle#L100-L138
[4] https://github.com/apache/beam/blob/master/sdks/python/apache_beam/io/gcp/bigquery_test.py#L625-L697

**Comment by pabloem** [comment]
@tvalentyn are you aware of why these tests could be getting triggered? See: https://github.com/apache/beam/pull/8871#issuecomment-511364580

**Comment by tvalentyn** [comment]
> @tvalentyn are you aware of why these tests could be getting triggered? See: [#8871 (comment)](https://github.com/apache/beam/pull/8871#issuecomment-511364580)

@pabloem, PTAL at https://issues.apache.org/jira/browse/BEAM-7674, it may be relevant to this question.

**Comment by ttanay** [comment]
Thanks @tvalentyn 
I made a change to that test. https://github.com/apache/beam/pull/9074

**Comment by ttanay** [comment]
Run Python PostCommit

**Comment by ttanay** [comment]
Run Python PostCommit

**Comment by ttanay** [comment]
Run Python PreCommit

**Comment by ttanay** [comment]
Run Python PostCommit

**Comment by ttanay** [comment]
Run Python PreCommit

**Comment by ttanay** [comment]
Run Python PostCommmit

**Comment by ttanay** [comment]
Run Python PostCommit

**Comment by ttanay** [comment]
Finally! :fireworks: 
All tests pass :heavy_check_mark: 
R: @pabloem

File: sdks/python/apache_beam/io/gcp/bigquery_file_loads.py [hunk 1]
```diff
@@ -622,8 +653,9 @@ def expand(self, pcoll):
                 test_client=self.test_client,
                 temporary_tables=self.temp_tables,
                 additional_bq_parameters=self.additional_bq_parameters),
-            load_job_name_pcv, *self.schema_side_inputs).with_outputs(
-                TriggerLoadJobs.TEMP_TABLES, main='main')
+            load_job_name_pcv, self.is_streaming_pipeline,
```
Comment (pabloem) [inline sdks/python/apache_beam/io/gcp/bigquery_file_loads.py]: I wonder if we should pass `is_streaming_pipeline` at pipeline construction (i.e. in the constructor) rather than a side input. It would allow us to show it as .p.ex. display data.

File: sdks/python/apache_beam/io/gcp/bigquery_file_loads.py [hunk 2]
```diff
@@ -550,6 +562,25 @@ def verify(self):
...
+
+  def _window_fn(self):
+    if self.is_streaming_pipeline:
+      return beam.WindowInto(beam.window.GlobalWindows(),
+                             trigger=trigger.Repeatedly(
+                                 trigger.AfterAny(
+                                     trigger.AfterProcessingTime(
+                                         self.triggering_frequency),
+                                     trigger.AfterCount(
+                                         _FILE_TRIGGERING_RECORD_COUNT))),
```
[the remark this answers is not in the document]
Comment (ttanay) [reply sdks/python/apache_beam/io/gcp/bigquery_file_loads.py]: I agree. With the current trigger, for large no. of files, the quota will be exceeded since multiple load jobs will be triggered. 
The Java implementation applies this trigger, writes the files and then triggers load jobs only based on the triggering frequency[1]. I now understand why that second trigger exists. I'll add it to this.

[1] https://github.com/apache/beam/blob/master/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BatchLoads.java#L272-L281

File: sdks/python/apache_beam/io/gcp/bigquery_file_loads.py [hunk 3]
```diff
@@ -550,6 +562,25 @@ def verify(self):
...
+                             accumulation_mode=trigger.AccumulationMode\
+                              .DISCARDING)
+    return beam.WindowInto(beam.window.GlobalWindows())
```
[the remark this answers is not in the document]
Comment (ttanay) [reply sdks/python/apache_beam/io/gcp/bigquery_file_loads.py]: Sure. I'll make the change.

File: sdks/python/apache_beam/io/gcp/bigquery_file_loads.py (hunk 1 above)
Comment (ttanay) [reply sdks/python/apache_beam/io/gcp/bigquery_file_loads.py]: According to the guide to generating Job IDs for BigQuery[1], the Job ID should be a human-readable prefix with a GUID/UUID suffix(Eg: `daily_import_job_1447971251`).
If we can generate the job ID as:
```python
"{}_{}".format(load_job_name_prefix, str(uuid.uuid4()))
```
 we can do away with passing the is_streaming_pipeline altogether. 
What are your thoughts? Will it make it harder for users to debug their jobs?

[1] https://cloud.google.com/bigquery/docs/running-jobs#generate-jobid

File: sdks/python/apache_beam/io/gcp/bigquery_file_loads.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (pabloem) [reply sdks/python/apache_beam/io/gcp/bigquery_file_loads.py]: Right, that trigger checks that there's at least one element, and then it waits the triggering frequency.

[the remark this answers is not in the document]
Comment (ttanay) [reply sdks/python/apache_beam/io/gcp/bigquery_file_loads.py]: Adding it..

**Comment by ttanay** [comment]
Run Python PostCommit

**Comment by ttanay** [comment]
Run Python 2 PostCommit

**Comment by ttanay** [comment]
I'm unable to run the IT tests for BQFL locally(even on clean master) and also on a VM due to this error:
```
ERROR: test_multiple_destinations_transform (apache_beam.io.gcp.bigquery_file_loads_test.BigQueryFileLoadsIT)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/tanay/Coding/beam-umbrella/beam-git/beam/sdks/python/apache_beam/io/gcp/bigquery_file_loads_test.py", line 467, in test_multiple_destinations_transform
    max_files_per_bundle=-1))
  File "/home/tanay/Coding/beam-umbrella/beam-git/beam/sdks/python/apache_beam/pipeline.py", line 426, in __exit__
    self.run().wait_until_finish()
  File "/home/tanay/Coding/beam-umbrella/beam-git/beam/sdks/python/apache_beam/pipeline.py", line 406, in run
    self._options).run(False)
  File "/home/tanay/Coding/beam-umbrella/beam-git/beam/sdks/python/apache_beam/pipeline.py", line 419, in run
    return self.runner.run_pipeline(self, self._options)
  File "/home/tanay/Coding/beam-umbrella/beam-git/beam/sdks/python/apache_beam/runners/direct/test_direct_runner.py", line 43, in run_pipeline
    self.result = super(TestDirectRunner, self).run_pipeline(pipeline, options)
  File "/home/tanay/Coding/beam-umbrella/beam-git/beam/sdks/python/apache_beam/runners/direct/direct_runner.py", line 128, in run_pipeline
    return runner.run_pipeline(pipeline, options)
  File "/home/tanay/Coding/beam-umbrella/beam-git/beam/sdks/python/apache_beam/runners/portability/fn_api_runner.py", line 319, in run_pipeline
    default_environment=self._default_environment))
  File "/home/tanay/Coding/beam-umbrella/beam-git/beam/sdks/python/apache_beam/runners/portability/fn_api_runner.py", line 326, in run_via_runner_api
    return self.run_stages(stage_context, stages)
  File "/home/tanay/Coding/beam-umbrella/beam-git/beam/sdks/python/apache_beam/runners/portability/fn_api_runner.py", line 408, in run_stages
    stage_context.safe_coders)
  File "/home/tanay/Coding/beam-umbrella/beam-git/beam/sdks/python/apache_beam/runners/portability/fn_api_runner.py", line 681, in _run_stage
    result, splits = bundle_manager.process_bundle(data_input, data_output)
  File "/home/tanay/Coding/beam-umbrella/beam-git/beam/sdks/python/apache_beam/runners/portability/fn_api_runner.py", line 1562, in process_bundle
    part_inputs):
  File "/usr/local/lib/python3.7/concurrent/futures/_base.py", line 586, in result_iterator
    yield fs.pop().result()
  File "/usr/local/lib/python3.7/concurrent/futures/_base.py", line 432, in result
    return self.__get_result()
  File "/usr/local/lib/python3.7/concurrent/futures/_base.py", line 384, in __get_result
    raise self._exception
  File "/usr/local/lib/python3.7/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/home/tanay/Coding/beam-umbrella/beam-git/beam/sdks/python/apache_beam/runners/portability/fn_api_runner.py", line 1561, in <lambda>
    self._registered).process_bundle(part, expected_outputs),
  File "/home/tanay/Coding/beam-umbrella/beam-git/beam/sdks/python/apache_beam/runners/portability/fn_api_runner.py", line 1500, in process_bundle
    result_future = self._controller.control_handler.push(process_bundle_req)
  File "/home/tanay/Coding/beam-umbrella/beam-git/beam/sdks/python/apache_beam/runners/portability/fn_api_runner.py", line 1017, in push
    response = self.worker.do_instruction(request)
  File "/home/tanay/Coding/beam-umbrella/beam-git/beam/sdks/python/apache_beam/runners/worker/sdk_worker.py", line 342, in do_instruction
    request.instruction_id)
  File "/home/tanay/Coding/beam-umbrella/beam-git/beam/sdks/python/apache_beam/runners/worker/sdk_worker.py", line 368, in process_bundle
    bundle_processor.process_bundle(instruction_id))
  File "/home/tanay/Coding/beam-umbrella/beam-git/beam/sdks/python/apache_beam/runners/worker/bundle_processor.py", line 593, in process_bundle
    data.ptransform_id].process_encoded(data.data)
  File "/home/tanay/Coding/beam-umbrella/beam-git/beam/sdks/python/apache_beam/runners/worker/bundle_processor.py", line 143, in process_encoded
    self.output(decoded_value)
  File "/home/tanay/Coding/beam-umbrella/beam-git/beam/sdks/python/apache_beam/runners/worker/operations.py", line 256, in output
    cython.cast(Receiver, self.receivers[output_index]).receive(windowed_value)
  File "/home/tanay/Coding/beam-umbrella/beam-git/beam/sdks/python/apache_beam/runners/worker/operations.py", line 143, in receive
    self.consumer.process(windowed_value)
  File "/home/tanay/Coding/beam-umbrella/beam-git/beam/sdks/python/apache_beam/runners/worker/operations.py", line 594, in process
    delayed_application = self.dofn_receiver.receive(o)
  File "/home/tanay/Coding/beam-umbrella/beam-git/beam/sdks/python/apache_beam/runners/common.py", line 795, in receive
    self.process(windowed_value)
  File "/home/tanay/Coding/beam-umbrella/beam-git/beam/sdks/python/apache_beam/runners/common.py", line 801, in process
    self._reraise_augmented(exn)
  File "/home/tanay/Coding/beam-umbrella/beam-git/beam/sdks/python/apache_beam/runners/common.py", line 868, in _reraise_augmented
    raise_with_traceback(new_exn)
  File "/home/tanay/Coding/beam-umbrella/beam-git/.env3/lib/python3.7/site-packages/future/utils/__init__.py", line 421, in raise_with_traceback
    raise exc.with_traceback(traceback)
  File "/home/tanay/Coding/beam-umbrella/beam-git/beam/sdks/python/apache_beam/runners/common.py", line 799, in process
    return self.do_fn_invoker.invoke_process(windowed_value)
  File "/home/tanay/Coding/beam-umbrella/beam-git/beam/sdks/python/apache_beam/runners/common.py", line 611, in invoke_process
    windowed_value, additional_args, additional_kwargs, output_processor)
  File "/home/tanay/Coding/beam-umbrella/beam-git/beam/sdks/python/apache_beam/runners/common.py", line 683, in _invoke_process_per_window
    windowed_value, self.process_method(*args_for_process))
  File "/home/tanay/Coding/beam-umbrella/beam-git/beam/sdks/python/apache_beam/runners/common.py", line 914, in process_outputs
    for result in results:
  File "/home/tanay/Coding/beam-umbrella/beam-git/beam/sdks/python/apache_beam/io/gcp/bigquery_file_loads.py", line 413, in process
    additional_load_parameters=additional_parameters)
  File "/home/tanay/Coding/beam-umbrella/beam-git/beam/sdks/python/apache_beam/io/gcp/bigquery_tools.py", line 594, in perform_load_job
    additional_load_parameters=additional_load_parameters)
  File "/home/tanay/Coding/beam-umbrella/beam-git/beam/sdks/python/apache_beam/utils/retry.py", line 197, in wrapper
    return fun(*args, **kwargs)
  File "/home/tanay/Coding/beam-umbrella/beam-git/beam/sdks/python/apache_beam/io/gcp/bigquery_tools.py", line 350, in _insert_load_job
    response = self.client.jobs.Insert(request)
  File "/home/tanay/Coding/beam-umbrella/beam-git/beam/sdks/python/apache_beam/io/gcp/internal/clients/bigquery/bigquery_v2_client.py", line 342, in Insert
    upload=upload, upload_config=upload_config)
  File "/home/tanay/Coding/beam-umbrella/beam-git/.env3/lib/python3.7/site-packages/apitools/base/py/base_api.py", line 731, in _RunMethod
    return self.ProcessHttpResponse(method_config, http_response, request)
  File "/home/tanay/Coding/beam-umbrella/beam-git/.env3/lib/python3.7/site-packages/apitools/base/py/base_api.py", line 737, in ProcessHttpResponse
    self.__ProcessHttpResponse(method_config, http_response, request))
  File "/home/tanay/Coding/beam-umbrella/beam-git/.env3/lib/python3.7/site-packages/apitools/base/py/base_api.py", line 604, in __ProcessHttpResponse
    http_response, method_config=method_config, request=request)
RuntimeError: apitools.base.py.exceptions.HttpBadRequestError: HttpError accessing <https://www.googleapis.com/bigquery/v2/projects//jobs?alt=json>: response: <{'vary': 'Origin, X-Origin, Referer', 'content-type': 'application/json; charset=UTF-8', 'date': 'Sat, 20 Jul 2019 17:08:07 GMT', 'server': 'ESF', 'cache-control': 'private', 'x-xss-protection': '0', 'x-frame-options': 'SAMEORIGIN', 'x-content-type-options': 'nosniff', 'alt-svc': 'quic=":443"; ma=2592000; v="46,43,39"', 'transfer-encoding': 'chunked', 'status': '400', 'content-length': '632', '-content-encoding': 'gzip'}>, content <{
  "error": {
    "code": 400,
    "message": "Invalid project ID ''. Project IDs must contain 6-63 lowercase letters, digits, or dashes. Some project IDs also include domain name separated by a colon. IDs must start with a letter and may not end with a dash.",
    "errors": [
      {
        "message": "Invalid project ID ''. Project IDs must contain 6-63 lowercase letters, digits, or dashes. Some project IDs also include domain name separated by a colon. IDs must start with a letter and may not end with a dash.",
        "domain": "global",
        "reason": "invalid"
      }
    ],
    "status": "INVALID_ARGUMENT"
  }
}
> [while running 'WriteWithMultipleDests/BigQueryBatchFileLoads/ParDo(TriggerLoadJobs)/ParDo(TriggerLoadJobs)']
-------------------- >> begin captured logging << --------------------
```
That's not the case with the tests being run in Jenkins. 

Trying to resolve this so I can debug the failing test.
Hopefully, it's just a problem with my installation.

**Comment by ttanay** [comment]
Run Python PostCommit

**Comment by ttanay** [comment]
Run Python 2 PostCommit

**Comment by ttanay** [comment]
Run Python 3.5 PostCommit

**Comment by ttanay** [comment]
Run Python 3.6 PostCommit

**Comment by ttanay** [comment]
Run Python 3.7 PostCommit

**Comment by ttanay** [comment]
Run Python 2 PostCommit

**Comment by ttanay** [comment]
Run Python 2 PostCommit

**Comment by ttanay** [comment]
Run Python 2 PostCommit

**Comment by ttanay** [comment]
The failure in Python PreCommit is die to task - `:sdks:python:docs`.
The failure in Python 2 PostCommit is due to task - `:sdks:python:crossLanguagePortableWordCount`.
All tests in task `:sdks:python:directRunnerIT` pass :ballot_box_with_check:

**Comment by ttanay** [comment]
Run Python 2 PostCommit

**Comment by ttanay** [comment]
Run Python 2 PostCommit

**Comment by ttanay** [comment]
Run Python_PVR_Flink PreCommit

**Comment by ttanay** [comment]
Run Python 2 PostCommit

**Comment by ttanay** [comment]
The same test - `apache_beam.io.gcp.bigquery_io_read_it_test.BigqueryIOReadIT.test_bigquery_read_1M_python` is failing on the master as well. The streaming tests in :sdks:python:directRunnerIT aren't being run due to it.

**Comment by ttanay** [comment]
Run Python 2 PostCommit

File: sdks/python/apache_beam/io/gcp/bigquery.py [hunk 4]
```diff
@@ -920,6 +920,7 @@ def __init__(self,
                additional_bq_parameters=None,
                table_side_inputs=None,
                schema_side_inputs=None,
+               triggering_frequency=None,
                validate=True):
     """Initialize a WriteToBigQuery transform.
```
Comment (ttanay) [inline sdks/python/apache_beam/io/gcp/bigquery.py]: Add corner cases as limitations.

**Comment by ttanay** [comment]
Run Python 2 PostCommit

**Comment by pabloem** [comment]
This looks good to me Tanay. The test looks good too. Let's add any pending changes, and ensure the tests pass - and we can merge.

**Comment by ttanay** [comment]
I've made the remaining changes:
 - Limitations comment
 - Change in load job names

All tests pass :heavy_check_mark: 
We can merge :)

File: sdks/python/apache_beam/io/gcp/bigquery_file_loads.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ttanay) [reply sdks/python/apache_beam/io/gcp/bigquery_file_loads.py]: Added

**Comment by ttanay** [comment]
Thank You @pabloem!
